### PR TITLE
DepthToSpace Operator Implementation

### PR DESCRIPTION
**Pull request by cagery**

### CODE REVIEW

**Comment by kahmed10** [comment]
Don't forget to disable the excludes in the onnx backend tests under `test/py/onnx_backend_test.py`.

**Comment by cagery** [comment]
> Don't forget to disable the excludes in the onnx backend tests under `test/py/onnx_backend_test.py`.

Thanks @kahmed10 ! Done.

File: src/onnx/parse_depthtospace.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            blocksize = info.attributes.at("blocksize").i();
+        }
+        // calculate dimensions
+        auto lens1 = s.lens();
+        auto lens2 = s.lens();
+        lens2[1]   = lens2[1] / std::pow(blocksize, 2);
+        lens2[2]   = lens2[2] * blocksize;
+        lens2[3]   = lens2[3] * blocksize;
+        if(mode == "DCR")
```
Comment (scxiao) [inline src/onnx/parse_depthtospace.cpp:38]: Can we rewrite the code something like the following?

```
        lens1.push_back(lens1[2]);
        lens1.push_back(lens1[3]);
        lens1[2] = blocksize;
        std::vector<int64_t> perm;
 	if(mode == "DCR")
        {
            lens1[1] = blocksize;
            lens1[3] = lens1[1] / std::pow(blocksize, 2);
            perm = {0, 3, 4, 1, 5, 2};
        }
        else if(mode == "CRD")
        {
            lens1[1] = lens1[1] / std::pow(blocksize, 2);
            lens1[3] = blocksize;
            perm = {0, 1, 4, 2, 5, 3};
        }
        else
            MIGRAPHX_THROW("DepthToSpace: mode attribute cannot be read.");

       auto temp1 = info.add_instruction(make_op("reshape", {{"dims", lens1}}), args[0]);
       auto temp2 = info.add_instruction(
                     make_op("transpose", {{"permutation", perm}}), temp1);
       return info.add_instruction(make_op("reshape", {{"dims", lens2}}),
                                info.make_contiguous(temp2));
```

File: test/onnx/onnx_test.cpp [hunk 2]
```diff
@@ -917,6 +917,51 @@ TEST_CASE(deconv_output_shape_3d_test)
     EXPECT(p == prog);
 }
 
+TEST_CASE(depthtospace_test)
```
Comment (scxiao) [inline test/onnx/onnx_test.cpp:920]: The three cases can be reodered according ini the alphabetical order.

File: src/onnx/parse_depthtospace.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            lens1.push_back(lens1[2]);
+            lens1.push_back(lens1[3]);
+            lens1[3] = lens1[1] / std::pow(blocksize, 2);
```
Comment (umangyadav) [inline src/onnx/parse_depthtospace.cpp:42]: Can we add check for the divisibility ? And also that `blockSize > 1`.

File: test/onnx/gen_onnx.py [hunk 4]
```diff
@@ -1016,6 +1016,51 @@ def deconv_stride_test():
     return ([node], [x, w], [y])
 
 
+@onnx_test
+def depthtospace_test():
+
+    x = helper.make_tensor_value_info('x', TensorProto.FLOAT, [2, 8, 5, 5])
```
Comment (umangyadav) [inline test/onnx/gen_onnx.py:1022]: Maybe use dtype other than float for one of the tests ?

**Review by umangyadav — CHANGES_REQUESTED** [review]
Looks good to me other than adding check for divisibility.

File: src/onnx/parse_depthtospace.cpp (hunk 1 above)
Comment (cagery) [reply src/onnx/parse_depthtospace.cpp:38]: Done, thanks @scxiao !

File: src/onnx/parse_depthtospace.cpp (hunk 3 above)
Comment (cagery) [reply src/onnx/parse_depthtospace.cpp]: Done

File: test/onnx/gen_onnx.py (hunk 4 above)
Comment (cagery) [reply test/onnx/gen_onnx.py:1022]: Not following, what's the benefit?

Comment (umangyadav) [reply test/onnx/gen_onnx.py:1022]: No benefit, but just  to assert it works with other dtypes as well. Up to you if you want to make change.

**Review by umangyadav — APPROVED** [review]

File: src/onnx/parse_depthtospace.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            blocksize = info.attributes.at("blocksize").i();
+        }
+        if(blocksize < 1)
+        {
+            MIGRAPHX_THROW("DepthToSpace: blocksize is less than 1");
+        }
+        // calculate dimensions
+        auto lens1 = s.lens();
+        auto lens2 = s.lens();
+        if((static_cast<int64_t>(lens2[1]) % static_cast<int64_t>(std::pow(blocksize, 2))) == 0)
```
Comment (pfultz2) [inline src/onnx/parse_depthtospace.cpp:39]: Why is the `static_cast` needed here? They aren't used for the division.

File: src/onnx/parse_depthtospace.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            MIGRAPHX_THROW("DepthToSpace: div by blocksize quotient not int ");
+        lens1.push_back(lens1[2]);
+        lens1.push_back(lens1[3]);
+        lens2[2] = lens2[2] * blocksize;
+        lens2[3] = lens2[3] * blocksize;
+        lens1[2] = blocksize;
+        std::vector<int64_t> perm;
+        if(mode == "DCR")
+        {
+            lens1[3] = lens1[1] / std::pow(blocksize, 2);
```
Comment (pfultz2) [inline src/onnx/parse_depthtospace.cpp:51]: Could we assign `std::pow(blocksize, 2)` to a local variable so its only calculated once?

**Review by scxiao — APPROVED** [review]

File: src/onnx/parse_depthtospace.cpp (hunk 6 above)
Comment (cagery) [reply src/onnx/parse_depthtospace.cpp]: Done!

File: src/onnx/parse_depthtospace.cpp (hunk 5 above)
Comment (cagery) [reply src/onnx/parse_depthtospace.cpp]: Makes sense, removed them.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**